File: scripts/benchmark.js
#!/usr/bin/env node

/**
 * FreeCAD CLI Benchmark
 *
 * Measures the performance of various CLI operations.
 */

const { spawn, execSync } = require('child_process');
const path = require('path');

const packageDir = path.resolve(__dirname, '..');

// ANSI colors
const colors = {
    reset: '\x1b[0m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    cyan: '\x1b[36m'
};

function log(message, color = 'reset') {
    console.log(`${colors[color]}${message}${colors.reset}`);
}

function runCommand(cmd, args) {
    const start = process.hrtime.bigint();

    try {
        execSync(`python -m freecad_cli ${args}`, {
            cwd: packageDir,
            encoding: 'utf8',
            timeout: 30000
        });
        const end = process.hrtime.bigint();
        const duration = Number(end - start) / 1e6; // ms
        return { success: true, duration };
    } catch (e) {
        const end = process.hrtime.bigint();
        const duration = Number(end - start) / 1e6;
        return { success: false, duration, error: e.message };
    }
}

async function benchmark() {
    log('\n🧪 FreeCAD CLI Benchmark', 'cyan');
    log('========================\n');

    const iterations = 100;

    // Test cases
    const tests = [
        {
            name: 'Part Creation (Box)',
            args: 'part create --name BenchBox --type Box --params \'{"length": 10, "width": 10, "height": 5}\''
        },
        {
            name: 'Sketch Creation',
            args: 'sketch create --name BenchSketch --plane XY'
        },
        {
            name: 'Draft Line',
            args: 'draft line --name BenchLine --x1 0 --y1 0 --x2 100 --y2 100'
        },
        {
            name: 'Document Info',
            args: 'document info'
        },
        {
            name: 'Object List',
            args: 'object list'
        },
        {
            name: 'JSON Help Output',
            args: '--format json part create --name Test --type Box'
        }
    ];

    const results = [];

    for (const test of tests) {
        log(`\n📊 Testing: ${test.name}`, 'blue');

        const times = [];

        for (let i = 0; i < iterations; i++) {
            const result = runCommand('python', ['-m', 'freecad_cli', ...test.args.split(' ')]);
            times.push(result.duration);

            if (!result.success) {
                log(`  ❌ Failed: ${result.error}`, 'yellow');
                break;
            }
        }

        if (times.length === iterations) {
            // Calculate statistics
            const avg = times.reduce((a, b) => a + b, 0) / times.length;
            const min = Math.min(...times);
            const max = Math.max(...times);

            // Calculate standard deviation
            const variance = times.reduce((sum, t) => sum + Math.pow(t - avg, 2), 0) / times.length;
            const stdDev = Math.sqrt(variance);

            log(`  ✅ Completed ${iterations} iterations`, 'green');
            log(`     Average: ${avg.toFixed(2)} ms`, 'reset');
            log(`     Min: ${min.toFixed(2)} ms`, 'reset');
            log(`     Max: ${max.toFixed(2)} ms`, 'reset');
            log(`     Std Dev: ${stdDev.toFixed(2)} ms`, 'reset');

            results.push({
                name: test.name,
                avg,
                min,
                max,
                stdDev
            });
        }
    }

    // Summary
    log('\n\n📈 Summary', 'cyan');
    log('==========\n');

    // Sort by average time
    results.sort((a, b) => a.avg - b.avg);

    log('Rank | Operation                    | Avg (ms) | Min (ms) | Max (ms) | StdDev');
    log('-----|------------------------------|----------|----------|----------|--------');

    results.forEach((r, i) => {
        log(
            `${String(i + 1).padStart(4)} | ${r.name.padEnd(28)} | ${r.avg.toFixed(2).padStart(8)} | ${r.min.toFixed(2).padStart(8)} | ${r.max.toFixed(2).padStart(8)} | ${r.stdDev.toFixed(2)}`,
            'reset'
        );
    });

    log('\n✅ Benchmark complete!\n');

    // Memory check
    const memUsage = process.memoryUsage();
    log(`Memory Usage:`, 'blue');
    log(`  RSS: ${(memUsage.rss / 1024 / 1024).toFixed(2)} MB`);
    log(`  Heap Used: ${(memUsage.heapUsed / 1024 / 1024).toFixed(2)} MB`);
    log(`  Heap Total: ${(memUsage.heapTotal / 1024 / 1024).toFixed(2)} MB\n`);
}

// Run benchmark
benchmark().catch(console.error);

File: benchmark.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
FreeCAD CLI Benchmark
====================

Measures the performance of FreeCAD CLI operations.
Can be run directly with Python without Node.js.

Usage:
    python benchmark.py
    python -m pytest tests/benchmark.py -v
"""

import time
import statistics
import sys
import os

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from freecad_cli.freecad_integration import FreeCADWrapper, get_wrapper


def benchmark_operation(func, name, iterations=100):
    """Benchmark a single operation"""
    print(f"\n📊 Testing: {name}")
    print("-" * 50)

    times = []
    wrapper = get_wrapper(headless=True)
    wrapper.initialize()

    for i in range(iterations):
        start = time.perf_counter()
        func(wrapper)
        end = time.perf_counter()
        times.append((end - start) * 1000)  # Convert to ms

    # Calculate statistics
    avg = statistics.mean(times)
    min_time = min(times)
    max_time = max(times)
    std_dev = statistics.stdev(times) if len(times) > 1 else 0

    print(f"  ✅ {iterations} iterations completed")
    print(f"     Average: {avg:.2f} ms")
    print(f"     Min: {min_time:.2f} ms")
    print(f"     Max: {max_time:.2f} ms")
    print(f"     Std Dev: {std_dev:.2f} ms")

    return {
        "name": name,
        "avg": avg,
        "min": min_time,
        "max": max_time,
        "std_dev": std_dev
    }


def run_benchmark():
    """Run all benchmarks"""
    print("\n" + "=" * 60)
    print("🧪 FreeCAD CLI Benchmark (Python)")
    print("=" * 60)

    iterations = 100

    # Test cases
    tests = [
        ("Part Creation (Box)", lambda w: w.create_part("Box", "Box", {"length": 10})),
        ("Part Creation (Cylinder)", lambda w: w.create_part("Cyl", "Cylinder", {"radius": 5})),
        ("Sketch Creation", lambda w: w.create_sketch("Sketch", "XY")),
        ("Sketch Line", lambda w: w.add_sketch_geometry("Sketch", "Line", {"x1": 0, "y1": 0, "x2": 10, "y2": 10})),
        ("Sketch Circle", lambda w: w.add_sketch_geometry("Sketch", "Circle", {"cx": 5, "cy": 5, "radius": 3})),
        ("Draft Line", lambda w: w.create_draft_object("Line", "Line", {"x1": 0, "y1": 0})),
        ("Draft Circle", lambda w: w.create_draft_object("Circle", "Circle", {"radius": 5})),
        ("Mesh Creation", lambda w: w.create_mesh_object("Mesh", "RegularMesh", {"width": 10, "height": 10})),
        ("Surface Creation", lambda w: w.create_surface("Surf", "Fill")),
        ("PartDesign Body", lambda w: w.create_partdesign_body("Body")),
        ("TechDraw Page", lambda w: w.techdraw_create_page("Page")),
        ("Spreadsheet Create", lambda w: w.spreadsheet_create("Sheet")),
        ("Assembly Create", lambda w: w.assembly_create("Assy")),
        ("FEM Analysis", lambda w: w.fem_create_analysis("Analysis")),
        ("Material Get Standard", lambda w: w.material_get_standard("Steel")),
        ("List Objects", lambda w: w.list_objects()),
        ("Get Object Info", lambda w: w.get_object_info("Test")),
        ("Boolean Fuse", lambda w: w.boolean_operation("Fuse", "Fuse", "Box1", "Box2")),
    ]

    results = []

    for name, func in tests:
        try:
            result = benchmark_operation(func, name, iterations)
            results.append(result)
        except Exception as e:
            print(f"  ❌ Failed: {e}")

    # Summary
    print("\n" + "=" * 60)
    print("📈 Summary")
    print("=" * 60)

    # Sort by average time
    results.sort(key=lambda x: x["avg"])

    print(f"\n{'Rank':<6} {'Operation':<30} {'Avg (ms)':<12} {'Min':<10} {'Max':<10}")
    print("-" * 70)

    for i, r in enumerate(results, 1):
        print(f"{i:<6} {r['name']:<30} {r['avg']:<12.2f} {r['min']:<10.2f} {r['max']:<10.2f}")

    print("\n" + "=" * 60)
    print("✅ Benchmark complete!")
    print("=" * 60 + "\n")


def run_single_operation_benchmark():
    """Run a quick benchmark of a single operation"""
    print("\n🔥 Quick Performance Test")
    print("-" * 40)

    iterations = 1000
    wrapper = get_wrapper(headless=True)
    wrapper.initialize()

    # Quick test
    start = time.perf_counter()
    for _ in range(iterations):
        wrapper.create_part("Box", "Box", {})
    end = time.perf_counter()

    total_time = (end - start) * 1000
    avg_time = total_time / iterations

    print(f"  {iterations} Part creations:")
    print(f"    Total: {total_time:.2f} ms")
    print(f"    Average: {avg_time:.4f} ms/op")
    print(f"    Throughput: {1000/avg_time:.0f} ops/sec")


if __name__ == "__main__":
    run_benchmark()
    run_single_operation_benchmark()
